plugin/healthchecker/filter.go

```diff
@@ -0,0 +1,115 @@
+package healthchecker
+
+import (
+	"time"
+
+	"github.com/bluele/gcache"
+	"github.com/miekg/dns"
+)
+
+type (
+	HealthCheckFilter struct {
+		cache    gcache.Cache
+		checker  Checker
+		interval time.Duration
+		names    map[string]struct{}
+	}
+
+	entry struct {
+		dnsRecord dns.RR
+		healthy   bool
+		quit      chan bool
+	}
+)
+
+func NewHealthCheckFilter(checker Checker, size, interval int, names map[string]struct{}) *HealthCheckFilter {
+	return &HealthCheckFilter{
+		cache: gcache.New(size).LRU().EvictedFunc(func(key interface{}, val interface{}) {
+			if e, ok := val.(entry); ok {
+				e.quit <- true
+			}
+		}).Build(),
+		checker:  checker,
+		interval: time.Duration(interval) * time.Millisecond,
+		names:    names,
+	}
+}
+
+func (p *HealthCheckFilter) FilterRecords(records []dns.RR) []dns.RR {
+	if p.cache.Len(false) == 0 {
+		recs := make([]dns.RR, len(records))
+		copy(recs, records)
+		go func() {
+			for _, r := range recs {
+				if _, ok := p.names[r.Header().Name]; ok || len(p.names) == 0 {
+					p.put(r)
+				}
+			}
+		}()
+		return records
+	}
+
+	result := make([]dns.RR, 0, len(records))
+
+	for _, r := range records {
+		if _, ok := p.names[r.Header().Name]; ok || len(p.names) == 0 {
```

Comment:
While `gcache` is goroutine safe, `p.names` is not.

And can you elaborate a bit the meaning of this map? Not quite sure I get it.

Reply:
This map is created on plugin setup and after that we just read from it (if I understand correctly)

Reply:
`p.names` is read-only, so we are safe, missed that.